### ISSUE TITLE
graph-tool --with-openmp --with-python3 fails referencing chrono namespace

### ISSUE THREAD

**Issue by gfairchild** [Homebrew/homebrew-science#5181]
### Indicate troubleshooting steps taken

- [X] Ran `brew update` and retried your prior step?
- [X] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?
- [X] Checked that a similar issue hasn't already been filed?
- [X] Looked up the software bug tracker if reporting a build error?
- [X] If you're seeing permission errors tried running `sudo chown -R $(whoami) $(brew --prefix)`?

### Bug reports / build errors:

```
> brew install graph-tool --with-openmp --with-python3
==> Installing graph-tool from homebrew/science
==> Using Homebrew-provided fortran compiler.
This may be changed by setting the FC environment variable.
==> Downloading https://downloads.skewed.de/graph-tool/graph-tool-2.20.tar.bz2
Already downloaded: /Users/gfairchild/Library/Caches/Homebrew/graph-tool-2.20.tar.bz2
==> ../configure --disable-debug --disable-dependency-tracking --prefix=/usr/local/Cellar/graph-tool/2.20 --enable-openmp PYTHON=python PYTHON_LDFLAGS=-undefined dynamic_lookup PYTHON_EXTRA_LDFLAGS=-undefined dynamic_lookup --with-python-module-path=/usr/local/Cellar/graph-tool/2.20/lib/python2.7/site-packages
==> make install
Last 15 lines from /Users/gfairchild/Library/Logs/Homebrew/graph-tool/02.make:
                 from /usr/local/include/boost/coroutine2/all.hpp:12,
                 from ../../../../src/graph/draw/../coroutine.hh:28,
                 from ../../../../src/graph/draw/graph_cairo_draw.cc:37:
/usr/local/include/boost/chrono/detail/is_evenly_divisible_by.hpp:17:11: note: candidates are: namespace boost::chrono { }
 namespace chrono {
           ^~~~~~
In file included from ../../../../src/graph/draw/graph_cairo_draw.cc:32:0:
/usr/local/Cellar/gcc/6.3.0_1/include/c++/6.3.0/chrono:59:13: note:                 namespace std::chrono { }
   namespace chrono
             ^~~~~~
make[3]: *** [graph_cairo_draw.lo] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [install-recursive] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1

READ THIS: http://docs.brew.sh/Troubleshooting.html

These open issues may also help:
graph-tool: symbol binding failed https://github.com/Homebrew/homebrew-science/issues/3259
graph-tool --with-python3 --with-openmp ignores python3, then fails https://github.com/Homebrew/homebrew-science/issues/4553
```

#### Brief summary of issue:

I'm trying to install [graph-tool](https://graph-tool.skewed.de/). I have all of the dependencies properly installed (as far as I can tell), but I'm getting the make error seen above. I've googled things related to graph-tool and chrono, but nothing related is popping up. I checked the [graph-tool issues](https://git.skewed.de/count0/graph-tool/issues) (open and closed), and nothing looks related (a search for "chrono", for example, yields no results).

#### The complete command that you issued:

`brew install graph-tool --with-openmp --with-python3`

#### Gist with output of `brew doctor`:

https://gist.github.com/gfairchild/ab9e7bd900ab1c28eb43920f4c3646b0

#### Link output by `brew gist-logs <formula>`:

https://gist.github.com/gfairchild/6ae146a676530db145e317d14fe6f5aa

#### Output of `brew config`:

```
> brew config
HOMEBREW_VERSION: 1.1.10
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 69d1cedd12d37d6a9535fbba17d2129b165c08f8
Last commit: 7 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c72390c035e99e3159fb95ce7483534dfad56d8d
Core tap last commit: 2 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: octa-core 64-bit haswell
Homebrew Ruby: 2.0.0-p648
Clang: 8.0 build 800
Git: 2.11.1 => /usr/local/bin/git
Perl: /usr/bin/perl
Python: /usr/local/bin/python => /usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/bin/ruby => /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby
Java: 1.8.0_121, 1.8.0_112, 1.8.0_102, 1.8.0_92, 1.8.0_77, 1.8.0_74, 1.8.0_72, 1.8.0_66, 1.8.0_60, 1.8.0_51, 1.8.0_40, 1.8.0_31, 1.7.0_71
macOS: 10.11.6-x86_64
Xcode: 8.2.1
CLT: 8.2.0.0.1.1480973914
X11: 2.7.9 => /opt/X11
```

**Comment by gfairchild** [comment]
I just corrected this particular error (but I encountered another, which I'll post shortly). Turns out the version of gcc that I installed was not installed using the `--without-multilib` flag. Installing gcc *without* passing this flag can make OpenMP break. All I had to do was this:

`brew reinstall gcc --without-multilib`

**Comment by iMichka** [comment]
@ilovezfs This does not directly depend on gcc6, but the dependencies install gcc, and this seems to be the compiler graph-tools picks up anyway.

Should we maybe add: `depends_on "gcc" => "--without-multilib"` ?

**Comment by gfairchild** [comment]
I think that could be a good idea. But that's only if the user attempts to build graph-tool with `--with-openmp`. I didn't try it, but my guess is that if I had tried to build graph-tool without OpenMP support that it would've worked.

**Comment by gfairchild** [comment]
I'm sorry, I seem to have spoken too soon. I thought that my issue was fixed, but that's because I actually ran `brew install graph-tool --with-python3` instead. I accidentally left off `--with-openmp`. Even after rebuilding gcc with `--without-multilib`, I'm still getting the same error I initially posted.

Any ideas what might be causing this?

**Comment by ilovezfs** [comment]
>Should we maybe add: depends_on "gcc" => "--without-multilib" ?

No, since the plan is to remove all `=> "with-foo"` at some point.

**Comment by iMichka** [comment]
This seems to compile with `boost@1.59`. I did not try the python build, just `--with-openmp`. It is still compiling on my computer (takes forever);. Could you try this too please?

**Comment by gfairchild** [comment]
I can do that tomorrow. I'm not on my work computer right now, but I'm almost positive I've got whatever the latest boost is (1.6x or something if I remember right). I hadn't considered a boost incompatibility, but that's a good thought. Please let me know if your build was successful.

**Comment by gfairchild** [comment]
Just checked, and I am using Boost 1.63.0, so it's possible that 1.6x is broken with graph-tool. About to attempt a Boost downgrade and rebuild.

**Comment by gfairchild** [comment]
Crap, it's not going to be easy to downgrade. I've got a number of installed packages that rely on the current boost.

```
> brew uninstall boost boost-python
Warning: Use numpy instead of deprecated homebrew/python/numpy
Error: Refusing to uninstall /usr/local/Cellar/boost/1.63.0
because it is required by curl 7.53.0, curl 7.53.1, nghttp2 1.18.1, nghttp2 1.19.0, nghttp2 1.20.0, osm2pgsql 0.92.0_3, postgis 2.3.2, which are currently installed.
You can override this and force removal with:
  brew uninstall --ignore-dependencies boost boost-python
```

Is there a way to install `boost@1.59` and tell graph-tool which version of Boost to use?

**Comment by iMichka** [comment]
You can replace the `depends_on "boost" => cxx11` with `depends_on "boost@1.59" => cxx11` in the formula.

I do not know if this fixes the python problem too. We will maybe need to move ` homebrew/versions/boost-python159` to `boost-python@1.59` to get this to work.

**Comment by ilovezfs** [comment]
It would be preferable to have upstream fix this.

**Comment by gfairchild** [comment]
Has anyone else tried building graph-tool with the current Boost 1.63.0? If so, are you getting the same compile error I am?

**Comment by geoglrb** [comment]
@gfairchild Yes, same error. For several versions of graph-tool, now.

**Comment by ilovezfs** [comment]
Closing since this is an upstream issue. Please report it to them.

**Comment by count0** [comment]
Hi, upstream here. I can't reproduce this problem with boost 1.63 (or any other boost version) in GNU/Linux. The fact that the error comes from deep within the Boost include files is rather strange.

Does it also happen when using clang?

**Comment by gfairchild** [comment]
I'm not sure, but I'm happy to try. However, I'm not sure I can tell the formula which compiler to use. These are the only flags:

```
> brew info graph-tool
homebrew/science/graph-tool: stable 2.22 (bottled), HEAD
efficient network analysis
https://graph-tool.skewed.de/
Not installed
From: https://github.com/Homebrew/homebrew-science/blob/master/graph-tool.rb
==> Dependencies
Build: pkg-config ✔
Required: boost ✔, boost-python ✔, cairomm ✔, cgal ✔, py2cairo ✔, gnome-icon-theme ✔, librsvg --with-gtk+3 ✔, pygobject3 ✔
Recommended: google-sparsehash ✔, gtk+3 ✔, numpy ✔, scipy ✔, matplotlib ✔
==> Requirements
Optional: python3 ✔
==> Options
--with-openmp
	Enable OpenMP multithreading
--with-python3
	Build with python3 support
--without-cairo
	Build without cairo support for plotting
--without-google-sparsehash
	Build without google-sparsehash support
--without-gtk+3
	Build without gtk+3 support for interactive plotting
--without-matplotlib
	Use a matplotlib you've installed yourself instead of a Homebrew-packaged matplotlib
--without-numpy
	Use a numpy you've installed yourself instead of a Homebrew-packaged numpy
--without-python
	Build without python2 support
--without-scipy
	Use a scipy you've installed yourself instead of a Homebrew-packaged scipy
--HEAD
	Install HEAD version
```

**Comment by gfairchild** [comment]
@ilovezfs, would you mind reopening this issue since upstream (@count0) is here?

**Comment by gfairchild** [comment]
@count0, have you tried building graph-tool using Homebrew, or are you just building from source?

**Comment by count0** [comment]
@gfairchild You can set the compiler globally in homebrew, I believe by setting the HOMEBREW_CXX environment variable.

I haven't build graph-tool with Homebrew, since I don't have a  MacOS machine.

**Comment by gfairchild** [comment]
Ah ha, so maybe this is part of the problem. Looks like I currently *am* using clang. I didn't even realize it.

```
> brew --env
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
SDKROOT: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk
MAKEFLAGS: -j8
CMAKE_PREFIX_PATH: /usr/local
CMAKE_INCLUDE_PATH: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/libxml2:/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include/apache2:/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
CMAKE_FRAMEWORK_PATH: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks
MACOSX_DEPLOYMENT_TARGET: 10.12
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/10.12
HOMEBREW_SDKROOT: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk
ACLOCAL_PATH: /usr/local/share/aclocal
PATH: /usr/local/Homebrew/Library/Homebrew/shims/super:/usr/bin:/bin:/usr/sbin:/sbin
```

So I need to try building with GCC. I wonder if that's the problem.

**Comment by count0** [comment]
Note how the compilation is using a GCC include file `/usr/local/Cellar/gcc/6.3.0_1/include/c++/6.3.0/chrono`. Something is inconsistent in your build environment.

**Comment by geoglrb** [comment]
Here's a case where I get it:

brew reinstall libsigc++ boost boost-python --build-from-source --cc=gcc-6
brew reinstall --build-from-source --with-openmp graph-tool

**Comment by geoglrb** [comment]
I have no problems with boost@159...

**Comment by count0** [comment]
@geoglrb Are you sure you are using the same compiler both for boost/boost-python and graph-tool?

**Comment by gfairchild** [comment]
@count0, after looking more closely at my [build logs](https://gist.github.com/gfairchild/6ae146a676530db145e317d14fe6f5aa), I am definitely using GCC, not clang. So the graph-tool formula must be overriding the compiler at some point.

**Comment by geoglrb** [comment]
It was my impression that graph-tool was using gcc-6 when I asked for --with-openmp, but I'll see if I can force that.

**Comment by geoglrb** [comment]
@count0 I do wonder why I can build with boost@159 but not 1.63. Almost the same configuration.

**Comment by count0** [comment]
There is obviously a conflict between boost::chrono and std::chrono, but I don't get it on my system with any boost version (including 1.63). I can change the source to make it unambiguous, but it is a bit puzzling.

**Comment by geoglrb** [comment]
What can I do to help you out in the most efficient manner? @count0

**Comment by count0** [comment]
You can try this patch: https://git.skewed.de/count0/graph-tool/commit/aaae585f18d324d50c3e1875b9b251c5c543d19e

**Comment by geoglrb** [comment]
Ok let me reproduce my own error and then directly apply.

**Comment by gfairchild** [comment]
I tried, but I'm getting another bug now that wasn't there before. I just recently updated to Sierra 10.12.3, so I'm not sure if that's the reason or if something else is going on:

```
> brew install graph-tool --with-openmp --HEAD
==> Installing graph-tool from homebrew/science
==> Using Homebrew-provided fortran compiler.
This may be changed by setting the FC environment variable.
==> Cloning https://git.skewed.de/count0/graph-tool.git
Updating /Users/gfairchild/Library/Caches/Homebrew/graph-tool--git
==> Checking out branch master
==> ./autogen.sh
==> ../configure --disable-debug --disable-dependency-tracking --prefix=/usr/local/Cellar/graph-tool/HEAD-aaae585 --enable-openmp PYTHON=python PYTHON_LDFLAGS=-undefined dynamic_lookup PYTHON_LIBS=-undefined dynamic_lookup PYTHON_EXTRA_LIBS=-undefined dynamic_lookup --with-python-module-path=/usr/local/Cellar/graph-tool/HEAD-aaae585/lib/python2.7/site-packages
Last 15 lines from /Users/gfairchild/Library/Logs/Homebrew/graph-tool/02.configure:
checking if g++-6 static flag -static works... no
checking if g++-6 supports -c -o file.o... yes
checking if g++-6 supports -c -o file.o... (cached) yes
checking whether the g++-6 linker (/usr/local/Homebrew/Library/Homebrew/shims/super/ld) supports shared libraries... yes
checking dynamic linker characteristics... darwin16.4.0 dyld
checking how to hardcode library paths into programs... immediate
checking for pkg-config... /usr/local/opt/pkg-config/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking whether g++-6 supports C++14 features by default... yes
checking whether to enable debug info... no
checking whether to enable parallel algorithms with openmp... yes
checking for OpenMP flag of C++ compiler... -fopenmp
checking whether to enable cairo drawing... yes
checking for Expat XML Parser... no
configure: error: expat library not found
```

I have expat installed:

```
> brew info expat
expat: stable 2.2.0 (bottled), HEAD [keg-only]
XML 1.0 parser
https://expat.sourceforge.io
/usr/local/Cellar/expat/2.2.0 (13 files, 381.2KB)
  Poured from bottle on 2017-03-07 at 10:07:50
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/expat.rb
==> Caveats
This formula is keg-only, which means it was not symlinked into /usr/local.

macOS includes Expat 1.5.

If you need to have this software first in your PATH run:
  echo 'export PATH="/usr/local/opt/expat/bin:$PATH"' >> ~/.bash_profile

For compilers to find this software you may need to set:
    LDFLAGS:  -L/usr/local/opt/expat/lib
    CPPFLAGS: -I/usr/local/opt/expat/include
For pkg-config to find this software you may need to set:
    PKG_CONFIG_PATH: /usr/local/opt/expat/lib/pkgconfig
```

So I'm not sure what's going on right now.

**Comment by count0** [comment]
@gfairchild You have to look inside config.log to see the reason of the failure.

**Comment by geoglrb** [comment]
@count0 The patch works for me!

Summary (without config or logs, which I would be happy to provide; each step involves proper modification of the homebrew formula not shown here):

     # this compiles using gcc-6 with openmp and works in python:
     brew reinstall libsigc++ boost@1.59 homebrew/versions/boost-python159 --build-from-source --cc=gcc-6 && brew reinstall --build-from-source --with-openmp graph-tool

     # unpatched, this fails to compile with the error of this issue:
      brew reinstall libsigc++ boost boost-python --build-from-source --cc=gcc-6 && brew reinstall --build-from-source --with-openmp --cc=gcc-6 graph-tool

     # patched with the above, this compiles and imports:
     brew reinstall libsigc++ boost boost-python --build-from-source --cc=gcc-6 && brew reinstall --build-from-source --with-openmp --cc=gcc-6 graph-tool

How else can I assist?

**Comment by gfairchild** [comment]
@geoglrb, the second two commands are identical. I think you meant to remove `--build-from-source` from the second part of the second command.

**Comment by geoglrb** [comment]
@gfairchild it's meant to be the same command. I just tried to make it clearer that the difference is applying a patch in the brew formula.

**Comment by gfairchild** [comment]
Ah, gotcha! 👍

**Comment by gfairchild** [comment]
I fixed my expat compilation issue by forcing Homebrew to link expat:

```
> brew link expat
Warning: expat is keg-only and must be linked with --force
Note that doing so can interfere with building software.
> brew link --force expat
Linking /usr/local/Cellar/expat/2.2.0... 9 symlinks created
```

I'm not sure why I need to do this in order to get this to compile. @geoglrb, did you encounter any expat strangeness when compiling? I have the standard bottle installed:

```
> brew info expat
expat: stable 2.2.0 (bottled), HEAD [keg-only]
XML 1.0 parser
https://expat.sourceforge.io
/usr/local/Cellar/expat/2.2.0 (13 files, 381.2KB) *
  Poured from bottle on 2017-03-07 at 13:39:04
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/expat.rb
==> Caveats
This formula is keg-only, which means it was not symlinked into /usr/local.

macOS includes Expat 1.5.

If you need to have this software first in your PATH run:
  echo 'export PATH="/usr/local/opt/expat/bin:$PATH"' >> ~/.bash_profile

For compilers to find this software you may need to set:
    LDFLAGS:  -L/usr/local/opt/expat/lib
    CPPFLAGS: -I/usr/local/opt/expat/include
For pkg-config to find this software you may need to set:
    PKG_CONFIG_PATH: /usr/local/opt/expat/lib/pkgconfig
```

I don't see why this would throw up an error. Perhaps I should open a new issue for this.

**Comment by count0** [comment]
@geoglrb I think we just need to add this patch to the formula, until the next version is released.

**Comment by geoglrb** [comment]
@gfairchild I don't know what the problem is with that. Looks similar to my expat when I 'brew info expat'. Sorry. I get these lines in the 01.configure:

     checking for OpenMP flag of C++ compiler... -fopenmp
     checking whether to enable cairo drawing... yes
     checking for Expat XML Parser headers in /usr/include... found
     checking for Expat XML Parser libraries... found
     checking for Expat XML Parser... yes

**Comment by geoglrb** [comment]
@count0 Great. I am preparing a pull request...

**Comment by iMichka** [comment]
The formula has been update with the patch. This should be good now. Thanks all for helping tackling this problem.
